refactor(parser): add from __future__ import annotations

ericbn opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #5013 

## Summary

### Changes

Add `from __future__ import annotations` to parser package

### User experience

Discussed in #4607

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [X] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [X] I have performed a self-review of this change
* [X] Changes have been tested
* [ ] Changes are documented
* [X] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

ericbn committed:
refactor(parser): add from __future__ import annotations

and update code according to ruff rules TCH, UP006, UP007, UP037 and
FA100.

ericbn commented on aws_lambda_powertools/utilities/parser/models/apigw.py:
```diff
@@ -1,10 +1,10 @@
+from datetime import datetime  # noqa: TCH003
+from typing import Any, Literal
 
-from aws_lambda_powertools.utilities.parser.types import Literal
```
I've changed all `from aws_lambda_powertools.utilities.parser.types import Literal` to directly import from typing to avoid the indirection internally.

leandrodamascena replied:
Perfect!

ericbn committed:
Fix type alias with Python 3.8

See https://bugs.python.org/issue45117

ericbn commented:
The code in this PR most probably fails in Python 3.8/3.9, like we saw with #4990.

EDIT: Added a new commit to fix it.

ericbn committed:
Fix pydantic not working with Python 3.8

    TypeError: You have a type annotation 'str | None' which makes use
    of newer typing features than are supported in your version of Python.
    To handle this error, you should either remove the use of new syntax
    or install the `eval_type_backport` package.

leandrodamascena committed:
Merge branch 'v3' into parser_annotations

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/apigw.py:
```diff
@@ -1,11 +1,10 @@
+# ruff: noqa: FA100
```
We don't need to add this line in every file, we can make it global in the `ruff.toml` file.

```
[lint.per-file-ignores]
# Ignore specific rules for specific files
"tests/e2e/utils/data_builder/__init__.py" = ["F401"]
"tests/e2e/utils/data_fetcher/__init__.py" = ["F401"]
"aws_lambda_powertools/utilities/data_classes/s3_event.py" = ["A003"]
"aws_lambda_powertools/utilities/parser/models/__init__.py" = ["E402"]
"aws_lambda_powertools/event_handler/openapi/compat.py" = ["F401"]
# Maintenance: we're keeping EphemeralMetrics code in case of Hyrum's law so we can quickly revert it
"aws_lambda_powertools/metrics/metrics.py" = ["ERA001"]
"aws_lambda_powertools/utilities/parser/models/*" = ["FA100"] <--- Add this line
```

ericbn replied:
Done

leandrodamascena reviewed (changes requested):
Hi @ericbn! I left a comment to improve the overall experience and make Ruff avoid `FA100` in all files under the `models` directory.

ericbn committed:
Update ruff.toml

Configure lint.per-file-ignores in ruff.toml instead of adding a

    # ruff: noqa: FA100

line to each file.

leandrodamascena committed:
Merge branch 'v3' into parser_annotations

leandrodamascena reviewed (approved):
APPROVED!
